subsys/instrumentation/Kconfig

```diff
@@ -0,0 +1,96 @@
+# Copyright 2023 Linaro
+# SPDX-License-Identifier: Apache-2.0
+
+# Instrumentation configuration options
+
+config INSTRUMENTATION
+	bool "Compiler Instrumentation Support"
+	imply REBOOT
+	imply RETAINED_MEM
+	imply RETAINED_MEM_ZEPHYR_RAM
+	imply RETENTION
+	imply THREAD_MONITOR
+	imply THREAD_NAME
+	imply THREAD_STACK_INFO
+	imply TIMING_FUNCTIONS
+	imply TRACING
+	imply UART_INTERRUPT_DRIVEN
+
+	help
+	  Enable compiler-managed runtime system instrumentation. This requires a
+	  compiler that supports function instrumentation, and a backend for
+	  capturing and processing those instrumentation events in the running
+	  application.
+
+if INSTRUMENTATION
+
+choice
+	prompt "Instrumentation Mode"
+	default INSTRUMENTATION_MODE_CALLGRAPH
+
+config INSTRUMENTATION_MODE_CALLGRAPH
+	bool "Call graph reconstruction (tracing)"
+	select THREAD_NAME
+	select INIT_STACKS
+	select THREAD_MONITOR
+	help
+	  Enables reconstruction of an instrumented call graph of function
+	  entry and exit events, within the limits of allocated memory.
+
+config INSTRUMENTATION_MODE_STATISTICAL
+	bool "Statistical (profiling)"
+	help
+	  Enables statistical profiling of the runtime system, tracking total
+	  execution time of the number of functions equal to TBD_VARIABLE.
+
+endchoice
+
+menu "Instrumentation Configuration"
+
+config INSTRUMENTATION_BUFFER_SIZE
+	int "Size of instrumentation buffer"
+	default 12000 if INSTRUMENTATION_MODE_CALLGRAPH
+	default 12000 if INSTRUMENTATION_MODE_STATISTICAL
+	range 20 65536 # Lowest value must match the biggest record size.
+	help
+	  Size of instrumentation buffer. If INSTRUMENTATION_MODE_CALLGRAPH is
+	  enabled, instrumentation buffer is used as a ring buffer to buffer entry
+	  and exit event records. If INSTRUMENTATION_MODE_STATISTICAL is enabled,
+	  each function encountered is added to the buffer for tracking until the
+	  buffer is full, at which point new functions will be ignored.
+
+config INSTRUMENTATION_BUFFER_OVERWRITE
+	bool "Instrumentation buffer overwriting mode"
+	depends on INSTRUMENTATION_MODE_CALLGRAPH
+	default y
+	help
+	  If instrumentation buffer is set as overwriting, then if the buffer
+	  gets full the older records will be overwritten by the newer records.
+	  This mode allows recording the most recent tracing events at the
+	  expense of losing the old ones. If this mode is not selected, then
+	  once the buffer is full tracing stops.
+
+config INSTRUMENTATION_SAVE_CONTEXT
+	bool "Save context on events"
+	default y
+	help
+	  If set, context info like CPU ID, thread priority, etc will be recorded
+	  on instrumenttion events, like on function entry / exit, when available.
```

Comment:
Totally agreed. In the early development, we thought it could be cases when the context could be ignored for tracing, reducing the size of the item recorded in the ringbuffer. However during the development I realized that doesn't make much sense. I've removed it now.